### ISSUE TITLE
Issue with initializing state with multiple symbolic arrays

### ISSUE THREAD

**Issue by lucaspena** [LeventErkok/sbv#681]
Hello,

I am having an issue with a simple example using multiple symbolic arrays:

```
import qualified Data.SBV as S

data St = St
  { f1 :: !Arr
  , f2 :: !Arr
  } deriving Show

type Arr = S.SArray Integer Integer

freshSt :: S.Symbolic St
freshSt = do
  field1 <- S.newArray "f1" Nothing
  field2 <- S.newArray "f2" Nothing
  return $ St { f1 = field1, f2 = field2 }

test :: S.Symbolic St
test = do
  st <- freshSt
  let newSt = st { f1 = S.writeArray (f1 st) 16 10 }
  let testOut = f1Test newSt 16 10
  P.liftIO $ do
    x <- testOut
    print x
  return st

f1Test :: St -> S.SInteger -> S.SInteger -> IO S.ThmResult
f1Test st k v = S.proveWith S.z3{S.verbose = True} $ S.readArray (f1 st) k S..== v
```
When I call `test`, the smt output looks as expected, and returns `Q.E.D`. However, if I swap the two initialization lines in `freshSt`
```
freshSt :: S.Symbolic St
freshSt = do
  field2 <- S.newArray "f2" Nothing
  field1 <- S.newArray "f1" Nothing
  return $ St { f1 = field1, f2 = field2 }
```
and call `test`, I get the following output:
```
** Calling: z3 -nw -in -smt2
[GOOD] ; Automatically generated by SBV. Do not edit.
[GOOD] (set-option :print-success true)
[GOOD] (set-option :global-declarations true)
[GOOD] (set-option :smtlib2_compliant true)
[GOOD] (set-option :diagnostic-output-channel "stdout")
[GOOD] (set-option :produce-models true)
[GOOD] (set-logic ALL) ; has unbounded values, using catch-all.
[GOOD] ; --- uninterpreted sorts ---
[GOOD] ; --- tuples ---
[GOOD] ; --- sums ---
[GOOD] ; --- literal constants ---
[GOOD] (define-fun s0 () Int 16)
[GOOD] (define-fun s1 () Int 10)
[GOOD] ; --- top level inputs ---
[GOOD] ; --- constant tables ---
[GOOD] ; --- non-constant tables ---
[GOOD] ; --- arrays ---
[GOOD] (declare-fun array_0 () (Array Int Int))
[FAIL] (define-fun array_0_initializer_0 () Bool (= array_0 (store array_1 s0 s1)))
draconic-runes-exe:
*** Data.SBV: Unexpected non-success response from Z3:
***
***    Sent      : (define-fun array_0_initializer_0 () Bool (= array_0 (store array_1 s0 s1)))
***    Expected  : success
***    Received  : (error "line 10 column 60: unknown constant array_1")
***
***    Exit code : ExitFailure (-15)
***    Executable: /Users/lucas.pena/.opam/default/bin/z3
***    Options   : -nw -in -smt2
***
***    Reason    : Check solver response for further information. If your code is correct,
***                please report this as an issue either with SBV or the solver itself!
```
Notably, the `[FAIL]` line seems to attempt to initialize `array_0` from `array_1` which doesn't exist yet. In the successful version, the corresponding line looks like
```
[GOOD] (define-fun array_0_initializer_0 () Bool (= array_0 (store array_0 s0 s1)))
```
It seems that reading/writing from the array that was not initialized first is causing problems. I don't understand why the order of initialization of the two arrays would be an issue. Please let me know if I am misunderstanding something or using symbolic arrays incorrectly or what. Thank you.

**Comment by LeventErkok** [comment]
Which version of SBV are you using? Can you confirm it's the latest? (i.e., 10.3 from https://hackage.haskell.org/package/sbv)

The fundamental problem here is that you're mixing and matching contexts, which is not allowed. Prior versions of SBV did not enforce context-consistency, but if you're using 10.3, you should get a proper error message.

Can you try with 10.3 and let me know if you still see an issue? (Note that you'll get an error message from SBV when you run it, but hopefully it should be self-explanatory. If not, let me know.)

**Comment by lucaspena** [comment]
Ah ok thanks, yea switching to 10.3 gets the mixing contexts error for each case. However, this arose from attempting to just read a (concrete) value from a symbolic array. First, I tried something like this:
```
main :: IO ()
main = S.runSMT readTest >>= print

readTest :: S.Symbolic (Maybe Integer)
readTest = do
  st <- freshSt
  let newSt = st { f1 = S.writeArray (f1 st) 16 10 }
  return $ readSym (f1 newSt) 16

readSym :: S.SArray Integer Integer -> S.SInteger -> Maybe Integer
readSym arr k = S.unliteral $ S.readArray arr k
```
but this prints `Nothing` where I would like it to print `Just 10`. Then I resorted to trying to "prove" the value was as expected as in my original post, but that led to the mixing contexts issue. Is there an easy way to just print out the value in a symbolic array? Provided a concrete key/value was written. Thanks, let me know.
